libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,333 @@
+selfFactor (S k) {ok} = CofactorExists 1 ok (rewrite multOneRightNeutral k in Refl)
+
+||| If b is a factor of a and c is a factor of b, then c is also a factor of a.
+factorTransitive : Factor a b -> Factor b c -> Factor a c
+factorTransitive (CofactorExists qb positA prfAB) (CofactorExists qc positB prfBC) =
+        CofactorExists (qb * qc) positA (
+            rewrite sym prfAB in
+            rewrite sym prfBC in
+            rewrite sym $ multAssociative c qc qb in
+            rewrite multCommutative qc qb in
+            Refl
+        )
+
+||| For all natural numbers p and q, p is a factor of (p * q).
+multFactor : (p, q : Nat) -> {auto positP : LTE 1 p} -> {auto positQ : LTE 1 q} -> Factor (p * q) p
+multFactor Z _ {positP} = absurd $ succNotLTEzero positP
+multFactor _ Z {positQ} = absurd $ succNotLTEzero positQ
+multFactor (S k) (S j) {positP} {positQ} =
+        CofactorExists (S j) (LTESucc LTEZero) Refl
+
+||| Any factor of n must be less than or equal to n.
+factorLteNumber : Factor n p -> LTE p n
+factorLteNumber (CofactorExists {n} {p} Z positN prf) =
+        let nIsZero = replace {P = \x => n = x} (multZeroRightZero p) $ sym prf
+            oneLteZero = replace {P = LTE 1} nIsZero positN
+        in
+        absurd $ succNotLTEzero oneLteZero
+factorLteNumber (CofactorExists {n} {p} (S k) positN prf) =
+        leftFactorLteProduct prf
+
+||| If p is a factor of n, then it is also a factor of (n + p).
+plusDivisorAlsoFactor : Factor n p -> Factor (n + p) p
+plusDivisorAlsoFactor (CofactorExists {n} {p} q positN prf) =
+        CofactorExists (S q) (lteTransitive positN $ lteAddRight n) $
+            rewrite plusCommutative n p in
+            rewrite multRightSuccPlus p q in
+            cong {f = plus p} prf
+
+||| If p is NOT a factor of n, then it also is NOT a factor of (n + p).
+plusDivisorNeitherFactor : NotFactor n p -> NotFactor (n + p) p
+plusDivisorNeitherFactor (ProperRemExists {n} {p} q r positN remPrf) =
+        ProperRemExists (S q) r (lteTransitive positN $ lteAddRight n) (
+                rewrite multRightSuccPlus p q in
+                rewrite sym $ plusAssociative p (p * q) (S $ finToNat r) in
+                rewrite plusCommutative p ((p * q) + S (finToNat r)) in
+                rewrite remPrf in
+                Refl
+            )
+
+||| If p is a factor of n, then it is also a factor of any multiply of n.
+multNAlsoFactor : Factor n p -> (a : Nat) -> {auto aok : LTE 1 a} -> Factor (n * a) p
+multNAlsoFactor _ Z {aok} = absurd $ succNotLTEzero aok
+multNAlsoFactor (CofactorExists {n} {p} q positN prf) (S a) =
+        CofactorExists (q * S a) (lteMultRight positN a) $
+            rewrite sym prf in
+            multAssociative p q (S a)
+
+||| If p is a factor of both n and m, then it is also a factor of their sum.
+plusFactor : Factor n p -> Factor m p -> Factor (n + m) p
+plusFactor {n} {p} nFactor@(CofactorExists qn positN prfN) (CofactorExists qm positM prfM) =
```

Comment:
What does the `@` mean?

Reply:
`@` binds the whole pattern on the right to the name on the left. In this case it binds `CofactorExists qn positN prfN` to the name `nFactor`. This way you can simultaneously refer to the whole expression and to its subexpressions.